aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -1112,6 +1112,24 @@ Tensor trace_cpu(const Tensor& self) {
   return result;
 }
 
+// TODO: this routine should be implemented without diag and sum for perf problems,
+// see https://github.com/pytorch/pytorch/pull/47305,
+Tensor linalg_trace(const Tensor& self, int64_t offset) {
+  TORCH_CHECK(self.dim() >= 2,
+           "self should have at least 2 dimensions, but has ", self.dim(), " dimensions instead");
+  TORCH_CHECK(-self.sizes()[0] < offset && offset < self.sizes()[1],
+           "offset is out of range [", -self.sizes()[0] + 1, ", ", self.sizes()[1] - 1, "]");
+  return at::sum(at::diagonal(self, offset, -2, -1), -1);
+}
+
+Tensor& linalg_trace_out(const Tensor& self, int64_t offset, Tensor &result) {
+  TORCH_CHECK(self.dim() >= 2,
+           "self should have at least 2 dimensions, but has ", self.dim(), " dimensions instead");
+  TORCH_CHECK(-self.sizes()[0] < offset && offset < self.sizes()[1],
```

Comment:
This should be `auto k = std::min(self.size(-2), self.size(-1));` and then use `k` in the check, right?

Reply:
I fixed the condition to `-self.size(-2) < offset && offset < self.size(-1)`. linalg.trace should accept `offset=9` for an input of size (1, 10), for example.

Reply:
You are completely right, sorry for that.